src/Netscript/killWorkerScript.ts

```diff
@@ -43,11 +43,9 @@ function stopAndCleanUpWorkerScript(ws: WorkerScript): void {
   ws.delayReject?.(new ScriptDeath(ws));
   ws.env.runningFn = "";
 
-  if (typeof ws.atExit === "function") {
+  for (const key in ws.atExit) {
     try {
-      const atExit = ws.atExit;
-      ws.atExit = undefined;
-      atExit();
+      ws.atExit[key]();
```

Comment:
You need to keep the same semantics here (empty the map immediately) to avoid infinite recursion.

Reply:
I dont quite see how this causes recursion, would you mind explaining?

regardless of that, instead of setting the key to undefined I think it would be better to straight up delete it then

```js
const atExit = ws.atExit[key];
delete ws.atExit[key];
atExit();
```

Reply:
nevermind, I think I just got it. If a script calls ns.exit inside ns.atExit that would cause the recursion right?

Reply:
the reason I want to delete instead of setting to undefined is that atExit is of type Record<string, () => void>, so typescript wouldnt catch that a key is undefined leading to the possibility of trying to call 'undefined' as function

Reply:
Dang, recursion really can get tricky!
Also yeah, I didn't think about performance cause I usually only work on projects where it really doesn't matter 😅

Ill make the changes you suggested and then test around a bit to make sure nothing is broken. Thanks for being so patient with this!

Reply:
I implemented the changes you suggested and also ran some ingame tests with the follwing script 
```js
/**
 * 
 * @param {NS} ns 
 */
export async function main(ns) {
  ns.atExit(() => {
    ns.tprint('AT EXIT PRE DEFAULT');
    ns.exit();
  }, 'PRE DEFAULT');

  ns.atExit(() => {
    ns.tprint('AT EXIT');
    ns.exit();
  });

  ns.atExit(() => {
    ns.tprint('AT EXIT POST DEFAULT');
    ns.exit();
  }, 'POST DEFAULT');

  ns.atExit(() => {
    ns.tprint('AT EXIT OVERRIDE TEST');
  }, 'OVERRIDE TEST');

  ns.atExit(() => {
    ns.tprint('OVERRIDDEN');
  }, 'OVERRIDE TEST');
}
```

This is the result 
![firefox_lSpHzy7KAO](https://github.com/bitburner-official/bitburner-src/assets/48560522/a426e9ba-1aad-49b4-9928-46f17a75c827)